### PR TITLE
refactor(serve): extract the preview server into its own module

### PR DESCRIPTION
**Pull request by yschimke**
#3824 preparation item 7 — the extraction itself. Follows #4597, which emptied `cliInternalsUsedByServe` and made this possible.

&gt; **This re-opens work that #4598 was supposed to land and didn't.** I pushed the wrong local branch name, so #4598's head was #4597's already-merged commit; it squash-merged to an **empty commit** (`353ea6ef1`) and none of these changes reached `main`. That commit is inert — `git diff 353ea6ef1^ 353ea6ef1` is empty — but its PR body describes changes that were never there. This PR is the same work, rebased onto `main` and verified to actually contain it (338 paths under `cli/serve/`).

`serve` was a package inside `:cli`, and a package has no boundary: the only thing keeping the server out of the CLI was `scripts/check-serve-seam.py`, a source scanner. It is **`:cli:serve`** now (`cli/serve`), 137 main sources plus tests and resources, and that direction is a build fact.

## `checkServeModuleBoundary`

The check #3824 asks for, now that there is a classpath to check. It walks the module's **resolved** runtime classpath — transitives included — and fails if `:cli`, a renderer, or the plugin appears.

Deliberately not "does `:cli` appear in my dependency block": that is a fact about the build file, which is exactly the thing a mistake would edit. A source scanner can also be defeated by reflection, a string literal, or a rule its tokenizer doesn't model; a resolved classpath cannot.

It matches on **resolved component identity** — `project :path` for a project in this build, `module group:name` for anything from a repository — rather than on where each classpath file sits on disk. The first version compared file paths against the forbidden projects' `projectDir`s and silently passed the case it most needed to catch: `renderers/desktop` publishes as `ee.schimke.composeai:renderer-desktop`, so once it arrives as a published or transitively substituted dependency its jar is in Gradle's cache, under no project directory at all. Falsified in both directions — with `implementation("ee.schimke.composeai:renderer-desktop:1.38.0")` on the server the path version reports `BUILD SUCCESSFUL` and the identity version fails naming the module; the project route fails naming the project.

That also fixed a promise the task had never kept: its description advertised the Gradle plugin as forbidden while no plugin entry existed in the map. `gradle-plugin` is an `includeBuild`, so `project(":gradle-plugin")` doesn't resolve from this build and it could never have been in a project-path list. It is covered now by published coordinates, the only route it can arrive by.

A direct `:cli` dependency is impossible anyway — Gradle rejects the cycle, which is the stronger guarantee. The `:cli` entry is there for transitive arrival. `:cli:check` depends on the task, which puts it inside `:cli:build` and therefore inside CI; wired only into `:cli:serve:check` it ran nowhere, since no CI job invokes a dependency project's `check` lifecycle.

## Package kept on purpose

Sources keep `ee.schimke.composeai.cli.serve`. Moving a module and renaming its package are independent changes — the lesson item 5 taught twice (see the design doc) — and doing both at once makes a 300-file diff unreviewable.

## Four things the move surfaced, none of which a compile would catch

- **`SERVE_VERSION` read `:cli`'s generated `cli-version.properties`**, which isn't on this module's classpath. Every `/version` read threw and the routing tests came back 500. The module generates its own `serve-version.properties` now — precisely the change `ServeVersion.kt`'s own comment (added in #4597) said this day would need.
- **Five test classes spelled repo paths as `File("../scripts/…")`**, correct only while the working directory was `cli/`. They now go through the `repoRoot()` helper that already sat beside them, which anchors on `settings.gradle.kts` instead of counting `..` segments.
- **`ServeWebAssetsTest` handed `node --check` the asset's own resource path.** That is `null` for a `jar:` URL, and whether the assets resolve from an exploded directory or from the module jar depends on packaging, not on anything the test asserts. It now checks the *bytes* — the more honest test, since it's the shipped copy that has to parse.
- **`stageRcFontResources` and the BTA / usage-PSI sidecar configurations** moved across with the code that needs them. Without the sidecars those tests don't fail — they take the fallback branch and pass, which is worse: the parser-backed rewrite and the in-process compile would have had no coverage while CI stayed green.

## Visibility cost: 4 symbols, not 102

I expected the module boundary to force ~102 `internal` symbols public. It didn't — 93 of the 102 `serveInternalsUsedByCli` entries were already implicitly public, and only **4** needed widening (`ServeBundleDaemon`, `ServeCatalogRefresher`, `ServeStartupBundles`, `openIsolatedSharedDaemonReplica`, plus `ServeCatalogChangeFeed`'s constructor).

`FakeRenderSession` becomes a **test fixture** rather than being duplicated, and `ServeRenderHost`'s session-wrapping *secondary* constructor becomes public while the daemon-spawning primary stays `internal` — wrapping a session the caller already owns is reasonable to expose; spawning one is not.

## Path pins, and why they are the expensive part of a move

A path pinned as a string literal outside Gradle is invisible to the compiler and to `:cli:check`, and several of these fail *quietly* rather than loudly — a missing player bundle skips, a missing scorer publishes references with no `match` results, a missing asset directory surfaces as ~90 unrelated-looking snapshot failures. Updated across `ci-paths.json`, `serve-lanes-paths.json`, `preview-scope.json`, six workflows, both allowlists, the two Playwright harnesses and `pages-snapshot.spec.mjs`, five `rc-*.test.mjs` suites, `usage-corpus.sh`, `compare-audit.mjs`, and ~40 documentation links. The asset directory now has an `existsSync` guard so the next move of this module is a one-line error instead of ninety red captures.

## What is left of item 7

`serveInternalsUsedByCli` is untouched at 102 on `main` — **92 of them `ServeCommand.kt` alone** (~4.9k lines, still in `:cli`). Reducing that to a thin entry point is the rest of the item, and is deliberately not attempted here.

## Test plan

Re-run after rebasing onto `main` at 1.39.0:

- `./gradlew :cli:serve:check :cli:compileTestKotlin` — green (5m38s). **2,410 tests** run in the new module, plus `checkServeModuleBoundary`.
- `./gradlew :cli:check` — green, including `checkServeSeam`, `checkCliDaemonLibraryBoundary`, `checkDaemonLaunchSchema` and the CLI's own suite.
- `./gradlew :cli:installDist` then `compose-preview --version` and `compose-preview serve --help` — both run from the assembled distribution, with `compose-preview-serve.jar` staged in `lib/`.
- `scripts/check-preview-server-contracts.sh` — green end to end.
- `preview-server/preview-harness` page captures — **173/173** locally (all `serve-*` snapshots and every `contract ·` test).
- `check-serve-seam.py` OK (`main.cliInternalsUsedByServe=0`); `test_check_serve_seam.py` 45 tests OK; `test_measure_serve_coupling.py` 42 OK; `check-daemon-launch-schema.py` OK; all five `.github/ci/test_*.py` suites OK.

Non-visual: a module boundary. No rendered output changes.

---
_Generated by [Claude Code](https://claude.ai/code/session_01YQJKxzuDLpRUepa8RYeaU5)_

### CODE REVIEW

File: scripts/check-serve-seam.py [hunk 1]
```diff
@@ -53,7 +53,20 @@ def cli_root(source_set: str) -> Path:
...
+    makes the diff unreviewable (see docs/design/PREVIEW_SERVER_SPLIT.md). So the path moved and the
+    package did not, which is why this is a separate root rather than `cli_root(...) / "serve"`.
+
+    The `serve → cli` direction this file used to police is now a *build* fact:
+    `:cli:serve` cannot depend on `:cli` because `:cli` depends on it, and
+    `checkServeModuleBoundary` fails on the resolved classpath if one ever arrives transitively.
+    What is left for this scanner is the `cli → serve` direction, which the build permits and the
+    split still needs to shrink.
+    """
+    return REPO_ROOT / "cli/serve/src" / source_set / "kotlin/ee/schimke/composeai/cli/serve"
```
[the remark this answers is not in the document]
Comment (yschimke) [reply scripts/check-serve-seam.py:69]: Right, and I falsified it rather than taking it on trust — this is the one of the six I'd have been most likely to wave through, because the gate passes either way.

With the old input declaration, appending a line to `cli/serve/src/main/.../ServeUrls.kt` and re-running gives:

```
> Task :cli:checkServeSeam UP-TO-DATE
```

With both trees declared, the same edit gives:

```
> Task :cli:checkServeSeam
check-serve-seam: OK — 122 crossing symbol(s), all on the allowlist …
```

So it wasn't "may report up-to-date" — it *did*, and the gate would have stopped running precisely where the coupling it polices now lives. Fixed in 762e4da6f by declaring both `cli/src` and `cli/serve/src`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cli/serve/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  )
+  add("composePreviewUsagePsi", project(":usage-source-psi"))
+}
+
+tasks.withType<Test>().configureEach {
+  // JUnit 5, as `:cli` runs these same tests today — `kotlin("test")` resolves its junit5 variant
+  // off the back of this, which is where the `org.junit.jupiter` API (`@TempDir`, `@Test`) comes
+  // from. Without it the platform defaults to JUnit 4 and roughly a dozen serve test classes stop
+  // compiling.
+  useJUnitPlatform()
```
[the remark this answers is not in the document]
Comment (yschimke) [reply cli/serve/build.gradle.kts:155]: Confirmed and fixed in 762e4da6f — both retargeted at `:cli:serve:test`.

There was a third piece you couldn't have seen from the diff alone: `UsageSnippetCorpusTest` reads `composeai.usageCorpus.repos` / `.out` / `.samples`, and the `providers.systemProperty(...)` forwarding that supplies them stayed behind in `:cli`'s test task when the test moved. So retargeting the script alone would have swapped one silent no-op for another — the test would have run and generated an empty corpus. The forwarding moved to `:cli:serve` with it, and `:cli`'s copy is removed since it now forwards for a test it no longer has.

`./gradlew :cli:serve:test --tests ee.schimke.composeai.cli.serve.ServeWebTest` runs green.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cli/serve/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+//
+// Same shape as `:cli`'s `generateCliVersionResource`, and the same value — both derive from
+// `project.version`, which honours the `PLUGIN_VERSION` env override CI sets and the
+// `.release-please-manifest.json` fallback for local builds. They are separate facts that happen
+// to agree today; when the server ships from its own repo they stop agreeing, and nothing here has
+// to change for that.
+val generateServeVersionResource =
+  tasks.register("generateServeVersionResource") {
+    val outputDir = layout.buildDirectory.dir("generated/serve-version-resource")
+    val serveVersion = project.version.toString()
```
[the remark this answers is not in the document]
Comment (yschimke) [reply cli/serve/build.gradle.kts:319]: Correct, and worse than "unspecified" reads: nothing type-checks a version string, so the server would have reported it through `/version`, the session handshake, page footers and every bug report without a single test noticing.

Verified before fixing — `./gradlew :cli:serve:properties` printed `version: unspecified`. Fixed in 762e4da6f with the same derivation `:cli` uses (PLUGIN_VERSION, else a patch-bumped SNAPSHOT off `.release-please-manifest.json`), and verified after: the generated resource now reads `version=1.39.1-SNAPSHOT`, and `compose-preview --version` from the assembled distribution agrees.

Deliberately a separate assignment rather than one shared with `:cli`: the two agree today because they ship together, and #3824's whole premise is that one day they won't — at which point the server reporting the server's version is the correct behaviour, not a bug to fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cli/serve/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+  implementation(project(":common-io"))
+  implementation(project(":data-layoutinspector-core"))
+  implementation(project(":data-theme-core"))
+  implementation(project(":data-fonts-core"))
+  implementation(project(":data-pseudolocale-core"))
+  implementation(project(":data-preview-overrides-core"))
+  implementation(project(":data-remotecompose-core"))
+  implementation(project(":data-render-core"))
+  implementation(project(":mcp"))
```
[the remark this answers is not in the document]
Comment (yschimke) [reply cli/serve/build.gradle.kts]: Verified and removed in 762e4da6f. The only `ee.schimke.composeai.mcp` occurrence under `cli/serve/src/main` is the KDoc link in `ServeOverrides.kt`, which needs no dependency to resolve at compile time — `:cli:serve:check` is green without it.

I had copied `:cli`'s dependency block wholesale when writing the module's build file, which is how it got there. Your framing is the right one: a module whose point is to be independently bounded shouldn't carry an MCP server on its runtime graph because the file it was cut from did. Kept on `:cli`, whose MCP commands import it for real.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/vscode-preview-baselines.yml [hunk 5]
```diff
@@ -15,7 +15,7 @@ on:
       # The serve fixture pages `<link>` the real viewer CSS/JS, which the harness server resolves
       # out of this directory — so a viewer change moves these captures. Only `format-compare.js`
       # was listed before, which left the pages themselves able to change without a rebaseline.
-      - 'cli/src/main/resources/ee/schimke/composeai/cli/serve/assets/**'
+      - 'cli/serve/src/main/resources/ee/schimke/composeai/cli/serve/assets/**'
```
[the remark this answers is not in the document]
Comment (yschimke) [reply .github/workflows/vscode-preview-baselines.yml:18]: Already fixed — you reviewed `64716ff7f9`, and this landed in 9fd0c9291 just before your review posted. All four consumers you named now resolve `cli/serve/src/main/resources/...`.

The finding was right, and the "fails soft" half is the part worth keeping on the record: `design-reference-score.mjs` returning no scorer and publishing references without `match` results is exactly the failure mode that makes these path pins dangerous. Same class caught `rc-player/bundle.js` and `lottie-player/`, which moved with the resources and are located by path in `ci.yml`, `design-artifacts-reusable.yml`, `ci-paths.json` and five `scripts/design-artifacts/rc-*.test.mjs` suites — a missing player bundle skips rather than errors, so those would have gone green testing nothing. All verified to resolve to real files.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/daemon-launch-schema-allowlist.json [hunk 6]
```diff
@@ -74,7 +74,7 @@
       "why": "`:render-session-subprocess` writes its own descriptors for the bundle daemons it spawns, stamping this private copy. It is one of the twelve published contract modules an extracted preview server links against (#3824), so post-split a stale mirror here is cross-repo version skew no compiler sees."
     },
     {
-      "file": "cli/src/main/kotlin/ee/schimke/composeai/cli/serve/ServeBundleDaemon.kt",
+      "file": "cli/serve/src/main/kotlin/ee/schimke/composeai/cli/serve/ServeBundleDaemon.kt",
```
[the remark this answers is not in the document]
Comment (yschimke) [reply scripts/daemon-launch-schema-allowlist.json:77]: Already fixed in ca3979e1c — you reviewed the head before it. `scripts/test_check_daemon_launch_schema.py:275` and `:317` now name the module path, and the suite is 55/55.

Diagnosis was exact, including which CI step it breaks. The instructive bit: I updated the allowlist for the move and not the test that reads it, so `check-daemon-launch-schema.py` itself reported **OK** while its own tests failed on the stale literal — a gate agreeing with a register that no longer described the tree.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cli/serve/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        ":daemon:android",
+        ":daemon:desktop",
+        ":renderer-android",
+        ":renderer-desktop",
+      )
+      .associateWith { project(it).projectDir.invariantSeparatorsPath }
+  )
+}
+
+tasks.named("check") { dependsOn("checkServeModuleBoundary") }
```
[the remark this answers is not in the document]
Comment (yschimke) [reply cli/serve/build.gradle.kts:301]: This is the most important finding on the PR, and you're right. Confirmed before fixing: `./gradlew :cli:build --dry-run` listed `checkServeModuleBoundary` **zero** times. The task this PR is named for — the thing that turns "serve cannot see the CLI" from a claim into a build fact — was reachable only by someone typing it by hand.

Fixed in 37afa6561: `:cli:check` now depends on it, which puts it inside `:cli:build`. Verified both directions:

- the dry-run graph now names `:cli:serve:checkServeModuleBoundary`;
- adding `implementation(project(":renderer-desktop"))` to the server makes `./gradlew :cli:check` fail with `Found on :cli:serve's resolved runtimeClasspath: :renderer-desktop (renderer-desktop-1.39.1-SNAPSHOT.jar)`.

Worth recording that this is the **third** gate this split produced that existed without running — after `checkServeSeam` going `UP-TO-DATE` on serve edits, and the daemon-launch gate agreeing with a register that no longer described the tree. A gate is two things, the rule and the wiring, and review naturally looks at the first. I'd written and falsified the rule here and never asked whether anything invoked it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cli/serve/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+  api(project(":bundle-coordinates"))
+  api(project(":daemon:core"))
+  api(project(":daemon-client"))
+  api(project(":render-session-api"))
+  api(project(":render-session-subprocess"))
+
+  implementation(project(":common-io"))
+  implementation(project(":data-layoutinspector-core"))
+  implementation(project(":data-theme-core"))
+  implementation(project(":data-fonts-core"))
```
[the remark this answers is not in the document]
Comment (yschimke) [reply cli/serve/build.gradle.kts]: Verified and removed in 37afa6561 — no source under `cli/serve/src` names `ee.schimke.composeai.data.fonts`, and `:cli:serve:check` (2410 tests) is green without it.

The floor argument is the one that decides it: `data-fonts-core` isn't among the probe's `contracts`, so carrying it made the server's real dependency floor wider than `checkContractSurface` asserts — a gap the probe couldn't have caught, since it measures the coordinates it's given rather than the module's actual graph.

Same root cause as the `:mcp` entry you flagged earlier: I wrote the module's build file by copying `:cli`'s dependency block and pruning what failed to compile, which keeps everything that is merely unnecessary. Compiling is the wrong test for a dependency you're trying not to have.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cli/serve/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+  @TaskAction
+  fun checkBoundary() {
+    val forbidden = forbiddenProjectDirs.get()
+    val hits =
+      runtimeClasspath.files
+        .mapNotNull { file ->
+          val path = file.invariantSeparatorsPath
+          forbidden.entries
+            .firstOrNull { (_, dir) -> path.startsWith("$dir/") }
+            ?.let { (project, _) -> "$project (${file.name})" }
```
[the remark this answers is not in the document]
Comment (yschimke) [reply cli/serve/build.gradle.kts]: Both halves confirmed and fixed in c82d365f1.

I falsified it in both directions rather than taking the reasoning on trust, because "a prefix compare misses a cached jar" is the kind of claim that sounds right and might still not be reachable here. It is reachable: `renderers/desktop` applies `composeai.maven-publishing` and publishes as `ee.schimke.composeai:renderer-desktop`, so your example is the literal case, not an analogy.

With `implementation("ee.schimke.composeai:renderer-desktop:1.38.0")` added to the server:

- old check → `BUILD SUCCESSFUL`
- new check → fails with `module ee.schimke.composeai:renderer-desktop`

and the project route still fails with `project :renderer-desktop`.

The plugin half is worse than "cannot be represented". The task's own description said it fails if "the CLI, a renderer, or the Gradle plugin" reaches the classpath, and there was no plugin entry in the map at all — `gradle-plugin` is an `includeBuild`, so `project(":gradle-plugin")` does not resolve from this build and never could have been in a project-path list. The description asserted a guarantee the rule did not implement, and nothing could have failed to reveal it. It is covered now by published coordinates, which is the only route it can actually arrive by.

The task now matches on resolved component identity — `project :path` for this build, `module group:name` for anything from a repository — with both forbidden sets named explicitly, since the same implementation has two identities depending on how it arrives. Config cache reusable; ktfmt clean.

That is four findings on this PR where the gate itself was the defect: `checkServeSeam` going `UP-TO-DATE`, the daemon-launch gate agreeing with a stale register, this boundary never being invoked, and now this boundary being unable to see its own headline case. The rule is the part review looks at; the wiring and the matching predicate are where these keep hiding.

---
_Generated by [Claude Code](https://claude.ai/code)_
